### PR TITLE
Refactor TLS

### PR DESCRIPTION
**Pull request by qinsoon**
This PR clarifies `tls` in mmtk-core. It addresses a task in https://github.com/mmtk/mmtk-core/issues/59.

This PR
* Introduces three new types, they replace the use of `OpaquePointer` in our code:
  * `VMThread`: any VM thread. We use them for allocators (which can be used by both mutators and collectors). 
  * `VMMutatorThread`: a VM thread with a `Mutator`
  * `VMWorkerThread`: a VM thread that runs a `GCWorker`
* We no longer needs unsafe for methods such as `mutator()`/`worker()`, as we know their types.
* reveals a potential bug in `MallocSpace` that we may poll from a collector.
* reveals a potential bug in reference processor where we try get a mutator from a worker thread and invoke its `flush_remember_set()`.

### CODE REVIEW

**Comment by qinsoon** [comment]
v8-binding-test
V8_BINDING_REF=refactor-tls

**Comment by qinsoon** [comment]
jikesrvm-binding-test
JIKESRVM_BINDING_REF=refactor-tls

**Comment by qinsoon** [comment]
openjdk-binding-test
OPENJDK_BINDING_REF=refactor-tls

File: src/plan/gencopy/mutator.rs [hunk 1]
```diff
@@ -6,17 +6,17 @@ use crate::plan::mutator_context::MutatorConfig;
...
 use enum_map::enum_map;
 use enum_map::EnumMap;
 
-pub fn gencopy_mutator_prepare<VM: VMBinding>(_mutator: &mut Mutator<VM>, _tls: OpaquePointer) {
+pub fn gencopy_mutator_prepare<VM: VMBinding>(_mutator: &mut Mutator<VM>, _tls: VMWorkerThread) {
     // Do nothing
 }
 
-pub fn gencopy_mutator_release<VM: VMBinding>(mutator: &mut Mutator<VM>, _tls: OpaquePointer) {
+pub fn gencopy_mutator_release<VM: VMBinding>(mutator: &mut Mutator<VM>, _tls: VMWorkerThread) {
```
[the remark this answers is not in the document]
Comment (qinsoon) [reply src/plan/gencopy/mutator.rs:19]: Sounds good. I will do it.

[the remark this answers is not in the document]
Comment (javadamiri) [reply src/plan/gencopy/mutator.rs:19]: I think there are many similar cases, but I'm not sure whether it's a good idea for you to spend time on adding all those comments.
Do it only if you think it'll take a reasonable amount of time

[the remark this answers is not in the document]
Comment (qinsoon) [reply src/plan/gencopy/mutator.rs:19]: We don't yet. I will create one.

[the remark this answers is not in the document]
Comment (javadamiri) [reply src/plan/gencopy/mutator.rs:19]: Thanks

**Review by javadamiri — APPROVED** [review]
LGTM

**Pull request merged**